developer/src/kmc-keyboard/src/kvk/kvks-file-reader.ts

```diff
@@ -0,0 +1,167 @@
+
+export default class KVKSFileReader {
+  private readonly callbacks: CompilerCallbacks;
+
+  constructor (callbacks: CompilerCallbacks) {
+    this.callbacks = callbacks;
+  }
+
+  public read(file: Uint8Array): VisualKeyboard {
+    let source = this.internalRead(file);
+    return this.transform(source);
+  }
+
+  public internalRead(file: Uint8Array): KVKSourceFile {
+    let source: KVKSourceFile;
+
+    const parser = new xml2js.Parser({
+      explicitArray: false,
+      mergeAttrs: true,
+      includeWhiteChars: true,
+      normalize: false,
+      emptyTag: {} as any
+      // Why "as any"? xml2js is broken:
+      // https://github.com/Leonidas-from-XIV/node-xml2js/issues/648 means
+      // that an old version of `emptyTag` is used which doesn't support
+      // functions, but DefinitelyTyped is requiring use of function or a
+      // string. See also notes at
+      // https://github.com/DefinitelyTyped/DefinitelyTyped/pull/59259#issuecomment-1254405470
+      // An alternative fix would be to pull xml2js directly from github
+      // rather than using the version tagged on npmjs.com.
+    });
+
+    parser.parseString(file, (e: unknown, r: unknown) => { source = r as KVKSourceFile });
+    source = this.boxArrays(source);
+    this.cleanupUnderscore('visualkeyboard', source.visualkeyboard);
+    return this.validate(source);
+  }
+
+  /**
+   * The only element that allows spaces is <key>. Remove
+   * all other empty whitespace-only values.
+   * @param root
+   * @param source
+   */
+  private cleanupUnderscore(root: string, source: any) {
+    if(root != 'key') {
+      if(source?.['_']?.trim() === '') {
+        delete source['_'];
+      }
+    }
+
+    for(let key of Object.keys(source)) {
+      if(Array.isArray(source[key])) {
+        for(let item of source[key]) {
+          if(typeof(item) === 'object') {
+            this.cleanupUnderscore(key, item);
+          }
+        }
+      } else if(typeof source[key] === 'object') {
+        this.cleanupUnderscore(key, source[key]);
+      }
+    }
+  }
+
+  public validate(source: KVKSourceFile): KVKSourceFile {
+    const schema = JSON.parse(this.callbacks.loadKvksJsonSchema().toString('utf8'));
+    const ajv = new Ajv();
+    if(!ajv.validate(schema, source)) {
+      console.dir(source, {depth:8});
+      this.callbacks.reportMessage(CompilerMessages.Error_InvalidFile({errorText: ajv.errorsText()}));
+      return null;
+    }
+    return source;
+  }
+
+  public transform(source: KVKSourceFile): VisualKeyboard {
+    // NOTE: at this point, the xml should have been validated
+    // and matched the schema result so we can assume properties exist
+    let result: VisualKeyboard = {
+      header: {
+        version: BUILDER_KVK_HEADER_VERSION,
+        flags: 0,
+        ansiFont: { name: "Arial", size: -12, color: 0xFF000008 }, // TODO-LDML: consider defaults
+        unicodeFont: { name: "Arial", size: -12, color: 0xFF000008 }, // TODO-LDML: consider defaults
```

Comment:
I wouldn't have guessed these defaults… yes, ideally they would be hoisted somewhere.

Reply:
I'll track this with those TODO-LDML comments